modules/http4s/test/src/smithy4s/http4s/NullsAndDefaultEncodingSuite.scala

```diff
@@ -0,0 +1,262 @@
+          )
+        ),
+        result
+      )
+    )
+  }
+
+  test("document encoder - default overrides") {
+    val result = Document.Encoder
+      .fromSchema(OperationOutput.schema)
+      .encode(
+        OperationOutput(
+          optional = Some("optional-override"),
+          optionalWithDefault = "optional-default-override",
+          requiredWithDefault = "required-default-override",
+          optionalHeader = Some("optional-header-override"),
+          optionalHeaderWithDefault = "optional-header-with-default-override",
+          requiredHeaderWithDefault = "required-header-with-default-override"
+        )
+      )
+    IO.pure(
+      assert.same(
+        Document.obj(
+          "optional" -> Document.fromString("optional-override"),
+          "optionalWithDefault" -> Document.fromString(
+            "optional-default-override"
+          ),
+          "requiredWithDefault" -> Document.fromString(
+            "required-default-override"
+          ),
+          "optionalHeader" -> Document.fromString(
+            "optional-header-override"
+          ),
+          "optionalHeaderWithDefault" -> Document.fromString(
+            "optional-header-with-default-override"
+          ),
+          "requiredHeaderWithDefault" -> Document.fromString(
+            "required-header-with-default-override"
+          )
+        ),
+        result
+      )
+    )
+  }
+
+  object Impl extends ServiceWithNullsAndDefaults[IO] {
+    override def operation(input: OperationInput): IO[OperationOutput] =
+      IO.pure(OperationOutput())
+  }
+
+  private val specHeaders = Set(
+    ci"optional-header",
+    ci"optional-header-with-default",
+    ci"required-header-with-default"
+  )
+
+  case class TestResponse(headers: Map[CIString, String], body: Json)
+
+  case class TestRequest(
+      headers: Map[CIString, String],
+      query: Map[String, String],
+      labels: List[String],
+      body: Json
+  )
+
+  private def runServerTest(explicitDefaults: Boolean): IO[TestResponse] = {
+    def run(
+        routes: HttpRoutes[IO],
+        req: Request[IO]
+    ): IO[(Map[CIString, String], Json)] =
+      routes.orNotFound.run(req).flatMap { response =>
+        response.as[Json].map(headersToMap(response.headers) -> _)
+      }
+    SimpleRestJsonBuilder
+      .withExplicitDefaultsEncoding(explicitDefaults)
+      .routes(Impl)
+      .resource
+      .use { routes =>
+        for {
+          result <- run(
+            routes,
+            Request[IO](method = Method.POST, uri = uri"/operation/label")
+          )
+          (headers, body) = result
+        } yield TestResponse(headers, body)
+      }
+  }
+
+  private def runClientTest(
+      explicitDefaults: Boolean,
+      input: OperationInput
+  ): IO[TestRequest] = {
+    val resources = for {
+      promise <- Deferred[IO, Request[IO]].toResource
+      httpClient: Client[IO] = Client(req =>
+        promise.complete(req).as(Response[IO]()).toResource
```

Comment:
**issue:** should mention that this doesn't necessarily copy over the entire request body. While we're not using the body, let's either use `req.toStrict` first, or leave a comment about it.